### PR TITLE
Git secret cat

### PR DESCRIPTION
**Pull request by joshrabinowitz**
For https://github.com/sobolevn/git-secret/issues/141

### CODE REVIEW

File: src/commands/git_secret_cat.sh [hunk 1]
```diff
@@ -0,0 +1,43 @@
+#!/usr/bin/env bash
+
+
+function cat {
+  local homedir=''
+  local passphrase=''
+
+  OPTIND=1
+
+  while getopts 'hfd:p:' opt; do
```
Comment (sobolevn) [inline src/commands/git_secret_cat.sh]: Do we really need `f` here?

File: src/commands/git_secret_cat.sh [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+  done
+
+  shift $((OPTIND-1))
+  [ "$1" = '--' ] && shift
+
+  _user_required
+
+  # Command logic:
+
+  #local path_mappings
```
Comment (sobolevn) [inline src/commands/git_secret_cat.sh]: These lines could easily be removed.

File: src/commands/git_secret_cat.sh [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+  #path_mappings=$(_get_secrets_dir_paths_mapping)
+
+  for line in "$@"
+  do
+    local filename
+    local path
+    #echo $line
```
Comment (sobolevn) [inline src/commands/git_secret_cat.sh]: This line should also be removed.

File: README.md [hunk 4]
```diff
@@ -73,18 +73,6 @@ Thank you to all our backers! 🙏 [[Become a backer](https://opencollective.com
 
 Support this project by becoming a sponsor. Your logo will show up here with a link to your website. [[Become a sponsor](https://opencollective.com/git-secret#sponsor)]
 
-<a href="https://opencollective.com/git-secret/sponsor/0/website" target="_blank"><img src="https://opencollective.com/git-secret/sponsor/0/avatar.svg"></a>
```
[the remark this answers is not in the document]
Comment (joshrabinowitz) [reply README.md]: It didn't seem like they were in the 0.2.3 release and also it didn't seem like the links work. Do you want me to put those lines back?

File: src/commands/git_secret_cat.sh (hunk 1 above)
Comment (joshrabinowitz) [reply src/commands/git_secret_cat.sh]: fixed

File: src/commands/git_secret_cat.sh (hunk 2 above)
Comment (joshrabinowitz) [reply src/commands/git_secret_cat.sh]: fixed

File: src/commands/git_secret_cat.sh (hunk 3 above)
Comment (joshrabinowitz) [reply src/commands/git_secret_cat.sh]: fixed

File: README.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (joshrabinowitz) [reply README.md]: Ok, sounds good. I have restored these lines.

File: tests/test_cat.bats [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+  set_state_git
+  set_state_secret_init
+  set_state_secret_tell "$TEST_DEFAULT_USER"
+  set_state_secret_add "$FILE_TO_HIDE" "$FILE_CONTENTS"
+  set_state_secret_hide
+}
+
+
+function teardown {
+  #rm "$FILE_TO_HIDE"
```
Comment (sobolevn) [inline tests/test_cat.bats]: please, remove this line

File: tests/test_cat.bats [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+
+  uninstall_fixture_full_key "$TEST_DEFAULT_USER" "$FINGERPRINT"
+  unset_current_state
+}
+
+
+@test "run 'cat' with password argument" {
+  cp "$FILE_TO_HIDE" "${FILE_TO_HIDE}2"
```
Comment (sobolevn) [inline tests/test_cat.bats]: Why do we need these two files here?
Can we just use one?

File: tests/test_cat.bats [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+
+  local password=$(test_user_password "$TEST_DEFAULT_USER")
+  run git secret cat -d "$TEST_GPG_HOMEDIR" -p "$password" "$FILE_TO_HIDE" 
+
+  [ "$status" -eq 0 ]
+
+  # $output is the output from 'git secret cat' above
+  [ "$FILE_CONTENTS" == "$output" ]
+
+  rm "${FILE_TO_HIDE}2"
```
Comment (sobolevn) [inline tests/test_cat.bats]: And if we will have just one file we would not have this line.

**Comment by sobolevn** [comment]
@joshrabinowitz you are awesome! 🌮

File: tests/test_cat.bats (hunk 5 above)
Comment (joshrabinowitz) [reply tests/test_cat.bats]: done

**Comment by joshrabinowitz** [comment]
> I would also personally ask for one more test: we should cover a situation when we try to show file which does not exist.

I added a test for

   git secret cat NO_SUCH_FILE

In that case the user should get an error indicating that the file was not found to decrypt, and returning an  exit code of 2.

File: tests/test_cat.bats (hunk 6 above)
Comment (joshrabinowitz) [reply tests/test_cat.bats]: fixed

File: tests/test_cat.bats (hunk 7 above)
Comment (joshrabinowitz) [reply tests/test_cat.bats]: fixed

File: tests/test_cat.bats [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+  run git secret cat -d "$TEST_GPG_HOMEDIR" -p "$password" "$FILE_TO_HIDE" 
+
+  [ "$status" -eq 0 ]
+
+  # $output is the output from 'git secret cat' above
+  # note that currently content may differ by a newline
+  [ "$FILE_CONTENTS" == "$output" ]
+
+  # these two lines needed for CI? Previous run stalled.
+  touch "$FILE_TO_HIDE)"
```
[the remark this answers is not in the document]
Comment (joshrabinowitz) [reply tests/test_cat.bats]: removed

File: tests/test_cat.bats [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+
+function teardown {
+  uninstall_fixture_full_key "$TEST_DEFAULT_USER" "$FINGERPRINT"
+  unset_current_state
+}
+
+
+@test "run 'cat' with password argument" {
+  # these two lines are needed for CI? Previous run stalled.
+  cp "$FILE_TO_HIDE" "${FILE_TO_HIDE}2"
```
[the remark this answers is not in the document]
Comment (joshrabinowitz) [reply tests/test_cat.bats]: removed

**Comment by joshrabinowitz** [comment]
OK, I think this is ready to go. 

Let me know if you'd like me to create a new PR with fewer commits for these changes.

Josh

**Comment by sobolevn** [comment]
Travis is broken.

I guess we have to do something with our CI, since it is so heavy right now.
Anyway, thank you @joshrabinowitz ! You did an awesome job.

**Comment by sobolevn** [comment]
@joshrabinowitz drop me a line if you would like to maintain the project!
#100

**Comment by joshrabinowitz** [comment]
Sure, I can help with maintenance. I'm thinking about the v0.2.4 release. How can I help.
Josh

**Comment by sobolevn** [comment]
@joshrabinowitz added you as a maintainer! Welcome onboard! 🎉

**Pull request merged**